### PR TITLE
SDI-192 Added visit outcome from dummy visitor record

### PR DESCRIPTION
**Pull request by andymarke**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponse.kt [hunk 1]
```diff
@@ -127,7 +127,8 @@ data class VisitResponse(
     visitorConcernText = visitEntity.visitorConcernText,
     visitors = visitEntity.visitors.filter { visitor -> visitor.person != null }
       .map { visitor -> Visitor(visitor.person!!.id, visitor.groupLeader) },
-    visitOutcome = null, // TODO
+    visitOutcome = visitEntity.visitors.find { it.person == null }?.outcomeReason
+      ?.let { CodeDescription(it.code, it.description) },
     leadVisitor = visitEntity.visitors.find { visitor -> visitor.groupLeader }?.person?.let {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponse.kt:132]: meat of change; visit outcome is held on the dummy visitor record

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Address.kt [hunk 2]
```diff
@@ -35,7 +35,7 @@ abstract class Address(
   @Column(name = "NO_FIXED_ADDRESS_FLAG")
   val noFixedAddressFlag: String = "N",
   @OneToMany(mappedBy = "address", cascade = [CascadeType.ALL], fetch = FetchType.LAZY, orphanRemoval = true)
-  @Where(clause = "OWNER_CLASS = '" + AddressPhone.PHONE_TYPE + "'")
+  @Where(clause = "OWNER_CLASS = '${AddressPhone.PHONE_TYPE}'")
   val phones: MutableList<AddressPhone> = ArrayList()
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Address.kt:39]: unrelated comment from previous PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Person.kt [hunk 3]
```diff
@@ -36,15 +36,15 @@ data class Person(
   val birthDate: LocalDate? = null,
 
   @OneToMany(mappedBy = "person", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
-  @Where(clause = "OWNER_CLASS = '" + PersonAddress.ADDR_TYPE + "'")
+  @Where(clause = "OWNER_CLASS = '${PersonAddress.ADDR_TYPE}'")
   val addresses: MutableList<PersonAddress> = mutableListOf(),
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Person.kt:40]: unrelated comment from previous PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/PersonAddress.kt [hunk 4]
```diff
@@ -18,7 +18,7 @@ class PersonAddress(
   locality: String? = null,
   startDate: LocalDate = LocalDate.now(),
   noFixedAddressFlag: String = "N",
-  phones: MutableList<AddressPhone> = ArrayList()
+  phones: MutableList<AddressPhone> = mutableListOf()
 ) : Address(
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/PersonAddress.kt:22]: unrelated comment from previous PR

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/VisitResourceIntTest.kt [hunk 5]
```diff
@@ -490,13 +491,6 @@ class VisitResourceIntTest : IntegrationTestBase() {
...
           .returnResult().responseBody!!
 
-        assertThat(visit.visitStatus.code).isEqualTo("SCH")
-        assertThat(visit.visitType.code).isEqualTo("SCON")
-        assertThat(visit.visitOutcome).isNull()
-        assertThat(visit.offenderNo).isEqualTo("A1234TT")
-        assertThat(visit.prisonId).isEqualTo("MDI")
-        assertThat(visit.startDateTime).isEqualTo(LocalDateTime.parse("2022-01-01T12:05"))
-        assertThat(visit.endDateTime).isEqualTo(LocalDateTime.parse("2022-01-01T13:05"))
         assertThat(visit.leadVisitor).isNotNull
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/VisitResourceIntTest.kt:494]: removed asserts already tested above when focus is on lead visitor

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Repository.kt [hunk 6]
```diff
@@ -69,7 +71,7 @@ class Repository(
             visit.visitors.addAll(
               visitBuilder.visitors.map {
                 it.build(it.person, leadVisitor = it.leadVisitor, visit)
-              }
+              } + listOf(visitBuilder.visitOutcome.build(visit, visitBuilder.visitOutcome.outcomeCode?.let { lookupVisitOutcomeReason(it) }))
             )
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Repository.kt:75]: in NOMIS the outcome is modelled as an extra dummy visitor, so we add it to the list of visitors

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponseTest.kt [hunk 7]
```diff
@@ -211,6 +227,34 @@ internal class VisitResponseTest {
...
+          visitors = mutableListOf(
+            visitor,
+            outcomeVisitorRecord.copy(outcomeReason = null)
+          )
+        )
+      val response = VisitResponse(theVisit)
+
+      assertThat(response.visitOutcome).isNull()
+    }
   }
```
Comment (lbennett-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponseTest.kt:258]: Do you need a test which has other records besides the dummy one?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponseTest.kt [hunk 8]
```diff
@@ -211,6 +227,34 @@ internal class VisitResponseTest {
...
 }
```
Comment (lbennett-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponseTest.kt:259]: hard to tell - but do you have a test with an outcome on a non visitor record just to check the dummy one is returned?

**Review by lbennett-moj — APPROVED** [review]

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponseTest.kt (hunk 8 above)
Comment (andymarke) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/VisitResponseTest.kt:259]: Good point - all the tests currently have no outcomes on the real visitor records where as think they would be there. I'll add a quick unit test to show they are ignored

**Pull request merged**